### PR TITLE
Migrate Validators from XCT to MAUI #41, #42, #40, #38, #37

### PR DESCRIPTION
**Pull request by VladislavAntonyuk**
## Fixes

#41, #42, #40, #39, #38, #37, #43

## Description

Migrate validators from XCT to MAUI

image
image

### CODE REVIEW

**Comment by VladislavAntonyuk** [comment]
@brminnick merge conflict are resolved

**Comment by brminnick** [comment]
Unblocking this now that .NET MAUI supports Unit Tests 🎉

File: src/CommunityToolkit.Maui.UnitTests/Behaviors/MultiValidationBehavior_Tests.cs [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+        // Arrange
+        var behavior = new CharactersValidationBehavior
+        {
+            CharacterType = characterType,
+            MinimumCharacterTypeCount = minimumCharactersNumber,
+            MaximumCharacterTypeCount = maximumCharactersNumber
+        };
+
+        var multiBehavior = new MultiValidationBehavior();
+        multiBehavior.Children.Add(behavior);
```
Comment (brminnick) [inline src/CommunityToolkit.Maui.UnitTests/Behaviors/MultiValidationBehavior_Tests.cs:49]: Let's add at least 3 `ValidationBehavior`s to `multiBehavior` to validate the "multi" part

**Review by brminnick — CHANGES_REQUESTED** [review]
Thanks Vlad! 

I added the missing EmailValidationBehavior_UnitTests.

But we still need a few things:
- RequiredString Sample
- TextValidationBehavior UnitTests
- TextValidationBehavior Sample
- ValidationBehavior Unit Tests
  - Include tests for all public Properties + Methods
  - I recommend creating `MockValidationBehavior`
    ```cs
    class MockValidationBehavior : ValidationBehavior
    ```

File: src/CommunityToolkit.Maui.UnitTests/Behaviors/EmailValidationBehavior_Tests.cs [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+			Text = value
+		};
+		entry.Behaviors.Add(behavior);
+
+		// Act
+		await behavior.ForceValidate();
+
+		// Assert
+		Assert.Equal(expectedValue, behavior.IsValid);
+	}
```
[the remark this answers is not in the document]
Comment (brminnick) [reply src/CommunityToolkit.Maui.UnitTests/Behaviors/EmailValidationBehavior_Tests.cs:67]: I got the test data from here: https://codefool.tumblr.com/post/15288874550/list-of-valid-and-invalid-email-addresses

**Comment by brminnick** [comment]
Hey @VladislavAntonyuk! Just a thought - if you split this PR into smaller PRs, we could get a few of the validators merged while we work on the features that still need updating.

File: src/CommunityToolkit.Maui/Behaviors/Validators/EmailValidationBehavior.shared.cs [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+
+namespace CommunityToolkit.Maui.Behaviors;
+
+/// <summary>
+/// The <see cref="EmailValidationBehavior"/> is a behavior that allows users to determine whether or not text input is a valid e-mail address. 
+/// For example, an <see cref="Entry"/> control can be styled differently depending on whether a valid or an invalid e-mail address is provided.
+/// The validation is achieved through a regular expression that is used to verify whether or not the text input is a valid e-mail address.
+/// It can be overridden to customize the validation through the properties it inherits from <see cref="Internals.ValidationBehavior"/>.
+/// </summary>
+public class EmailValidationBehavior : TextValidationBehavior
```
[the remark this answers is not in the document]
Comment (VladislavAntonyuk) [reply src/CommunityToolkit.Maui/Behaviors/Validators/EmailValidationBehavior.shared.cs:14]: to support all test cases we have to use this regex: http://www.ex-parrot.com/pdw/Mail-RFC822-Address.html (the link from the website you suggested https://codefool.tumblr.com/post/15288874550/list-of-valid-and-invalid-email-addresses).
I think the regex looks ugly and almost impossible to maintain.
I am voting to MailAddress.TryCreate. if somebody creates a bug for the email validator, we will solve it for the specific scenario (maybe using regex)

File: src/CommunityToolkit.Maui.UnitTests/Behaviors/EmailValidationBehavior_Tests.cs (hunk 2 above)
[the remark this answers is not in the document]
Comment (VladislavAntonyuk) [reply src/CommunityToolkit.Maui.UnitTests/Behaviors/EmailValidationBehavior_Tests.cs]: the regex looks ugly for these scenarios: http://www.ex-parrot.com/pdw/Mail-RFC822-Address.html
even more, not all of the specified scenarios are valid. I saw this email `email@example` and it worked like a charm.

**Comment by VladislavAntonyuk** [comment]
> Hey @VladislavAntonyuk! Just a thought - if you split this PR into smaller PRs, we could get a few of the validators merged while we work on the features that still need updating.

You are right. I thought the first PR would be the exact copy of XCT.
This PR requires base ValidationBehavior tests, so we cannot merge it.
I have almost finished the implementation. The only ValidationBehaviour left. So I think you can review and if everything looks good merge it tomorrow.

**Review by brminnick — APPROVED** [review]

**Pull request merged**